pkg/grpc/proxy/handler.go

```diff
@@ -46,70 +53,98 @@ func RegisterService(server *grpc.Server, director StreamDirector, serviceName s
 // backends. It should be used as a `grpc.UnknownServiceHandler`.
 //
 // This can *only* be used if the `server` also uses grpcproxy.CodecForServer() ServerOption.
-func TransparentHandler(director StreamDirector) grpc.StreamHandler {
-	streamer := &handler{director}
+func TransparentHandler(director StreamDirector, resiliency resiliency.Provider) grpc.StreamHandler {
+	streamer := &handler{
+		director:   director,
+		resiliency: resiliency,
+	}
 	return streamer.handler
 }
 
 type handler struct {
-	director StreamDirector
+	director      StreamDirector
+	resiliency    resiliency.Provider
+	bufferedCalls []interface{}
 }
 
 // handler is where the real magic of proxying happens.
 // It is invoked like any gRPC server stream and uses the gRPC server framing to get and receive bytes from the wire,
 // forwarding it to a ClientStream established against the relevant ClientConn.
 func (s *handler) handler(srv interface{}, serverStream grpc.ServerStream) error {
+	// Clear the buffered calls on a new request.
+	s.bufferedCalls = []interface{}{}
+
 	// little bit of gRPC internals never hurt anyone
 	fullMethodName, ok := grpc.MethodFromServerStream(serverStream)
 	if !ok {
 		return status.Errorf(codes.Internal, "lowLevelServerStream not exists in context")
 	}
-	// We require that the director's returned context inherits from the serverStream.Context().
-	outgoingCtx, backendConn, err := s.director(serverStream.Context(), fullMethodName)
-	if err != nil {
-		return err
-	}
 
-	clientCtx, clientCancel := context.WithCancel(outgoingCtx)
-	// TODO(mwitkow): Add a `forwarded` header to metadata, https://en.wikipedia.org/wiki/X-Forwarded-For.
-	clientStream, err := grpc.NewClientStream(clientCtx, clientStreamDescForProxying, backendConn, fullMethodName, grpc.CallContentSubtype((&codec.Proxy{}).Name()))
-	if err != nil {
-		return err
+	// Fetch the AppId so we can reference it for resiliency.
+	md, _ := metadata.FromIncomingContext(serverStream.Context())
+	v := md.Get(diagnostics.GRPCProxyAppIDKey)
```

Comment:
Can we move that to the section where a resiliency policy is used?

Reply:
We need it to get the resiliency policy, so it has to be before the policy. Or am I misunderstanding what you're asking?

Reply:
Yeah I realize it's needed for the policy.